### PR TITLE
Mm 42153 update daily digest message

### PR DESCRIPTION
**Pull request by gbochora**
#### Summary
This PR updates the daily digest and `/playbook todo` messages contents.

- The daily digest now only shows tasks that are overdue or due today and tasks without a due date. For tasks that are due after today, just summary information is displayed:
`You have **x assigned tasks due after today**. Please use /playbook todo to see all your tasks.`

- `/playbook todo` message now includes information about tasks due date.

**`/playbook todo` message** screenshot:
Screenshot 2022-03-31 at 18 54 07

**Daily digest** screenshot:
Screenshot 2022-03-31 at 18 53 45

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42153

#### Checklist

- ~~[ ] Telemetry updated~~
- ~~[ ] Gated by experimental feature flag~~
- [x] Unit tests updated

### CODE REVIEW

File: server/app/playbook_run_service_test.go [hunk 1]
```diff
@@ -1306,3 +1308,327 @@ func TestMultipleWebhooks(t *testing.T) {
 		}
 	})
 }
+
+func TestDMTodoDigestToUser(t *testing.T) {
+	controller := gomock.NewController(t)
```
Comment (crspeller) [inline server/app/playbook_run_service_test.go:1313]: That's a lot of mocking just to get this test. Is there a way we could do this as a test in `server/api_runs_test.go`?

File: server/api/bot.go [hunk 2]
```diff
@@ -216,7 +216,7 @@ func (h *BotHandler) connect(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
-		if err = h.playbookRunService.DMTodoDigestToUser(userID, false); err != nil {
+		if err = h.playbookRunService.DMTodoDigestToUser(userID, false, timezone); err != nil {
```
Comment (crspeller) [inline server/api/bot.go:219]: Don't we want to always use the TZ behavior of the user object? Why do we prefer the browser one?

**Review by crspeller — APPROVED** [review]
Couple of non-blocking suggestions.

File: server/api/bot.go (hunk 2 above)
Comment (gbochora) [reply server/api/bot.go]: You are right. I first implemented the `daily digest` part and didn't know about the existence of user TZ at that point.

File: server/app/playbook_run_service.go [hunk 3]
```diff
@@ -2625,31 +2632,93 @@ func triggerWebhooks(s *PlaybookRunServiceImpl, webhooks []string, body []byte)
 
 }
 
-func buildAssignedTaskMessageAndTotal(runs []AssignedRun, locale string) string {
+func buildAssignedTaskMessageSummery(runs []AssignedRun, locale string, timezone *time.Location, onlyDueUntilToday bool) string {
```
Comment (agarciamontoro) [inline server/app/playbook_run_service.go:2635]: Typo
```suggestion
func buildAssignedTaskMessageSummary(runs []AssignedRun, locale string, timezone *time.Location, onlyDueUntilToday bool) string {
```

File: server/app/playbook_run_service.go [hunk 4]
```diff
@@ -2625,31 +2632,93 @@ func triggerWebhooks(s *PlaybookRunServiceImpl, webhooks []string, body []byte)
...
-		return msg + T("app.user.digest.tasks.zero_outstanding") + "\n"
+		msg.WriteString(T("app.user.digest.tasks.zero_assigned"))
+		msg.WriteString("\n")
+		return msg.String()
 	}
 
-	msg += T("app.user.digest.tasks.num_outstanding", total) + "\n\n"
+	var tasksNoDueDate, tasksDoAfterToday int
+	now := model.GetMillis()
+	currentTime := time.Unix(now/1000, 0).In(timezone)
```
Comment (agarciamontoro) [inline server/app/playbook_run_service.go:2656]: There is a function in `timeutils` you can use for this type of conversion: `timeutils.GetTimeForMillis`: https://github.com/mattermost/mattermost-plugin-playbooks/blob/master/server/timeutils/timeutils.go#L9-L11

File: server/app/playbook_run_service.go [hunk 5]
```diff
@@ -2625,31 +2632,93 @@ func triggerWebhooks(s *PlaybookRunServiceImpl, webhooks []string, body []byte)
...
+		if onlyDueUntilToday {
+			msg.WriteString(T("app.user.digest.tasks.num_assigned_due_until_today", total-tasksDoAfterToday))
+		} else {
+			msg.WriteString(T("app.user.digest.tasks.num_assigned", total))
+		}
+		msg.WriteString("\n\n")
+		msg.WriteString(runsInfo.String())
+	}
+
+	// add summery info for tasks without a due date or due date after today
```
Comment (agarciamontoro) [inline server/app/playbook_run_service.go:2714]: Typo
```suggestion
	// add summary info for tasks without a due date or due date after today
```

File: server/app/playbook_run_service_test.go (hunk 1 above)
Comment (gbochora) [reply server/app/playbook_run_service_test.go:1313]: The daily digest message doesn't have an API endpoint, so I think there is no way.

`/playbook todo` has an API endpoint, but it doesn't return post text. Instead, it just posts a message in the channel, so we must retrieve text from the database.

File: server/app/playbook_run_service.go [hunk 6]
```diff
@@ -2772,3 +2841,25 @@ func (s *PlaybookRunServiceImpl) dmPostToUsersWithPermission(users []string, pos
...
+
+// getDaysDiff returns days difference between two date.
+func getDaysDiff(start, end time.Time) int {
+	days := int(end.Sub(start).Hours() / 24)
+
+	if start.AddDate(0, 0, days).YearDay() != end.YearDay() {
+		days++
+	}
+	return days
+}
```
Comment (agarciamontoro) [inline server/app/playbook_run_service.go:2865]: What do you think about moving these utils functions to `timeutils` too? I think they may be useful elsewhere

File: server/app/playbook_run_service_test.go [hunk 7]
```diff
@@ -1306,3 +1308,326 @@ func TestMultipleWebhooks(t *testing.T) {
...
+		store.EXPECT().GetRunsWithAssignedTasks(user.Id).Return(assignedRuns, nil)
+		var digestPost *model.Post
+		poster := mock_bot.NewMockPoster(controller)
+		// catch post object passed to DM function to check text
+		poster.EXPECT().DM(user.Id, gomock.Any()).DoAndReturn(
+			func(userID string, post *model.Post) interface{} {
+				digestPost = post
+				return nil
+			},
+		)
```
Comment (agarciamontoro) [inline server/app/playbook_run_service_test.go:1465]: I did not know about `DoAndReturn`, that's cool!

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Looks good! Coding this type of logic is always cumbersome, but the result is great :) I just found a couple of typos and I left a suggestion about moving some code, but nothing important.

File: server/app/playbook_run_service.go (hunk 6 above)
Comment (gbochora) [reply server/app/playbook_run_service.go:2865]: Excellent idea! Didn't know about `timeutils`.

**Review by agarciamontoro — APPROVED** [review]
:tada:

**Review by abhijit-singh — APPROVED** [review]
Looks good!

**Pull request merged**